exec/bigmachine.go

```diff
@@ -170,43 +175,55 @@ func (b *bigmachineExecutor) manager(i int) *machineManager {
 
 type invocationRef struct{ Index uint64 }
 
-func (b *bigmachineExecutor) compile(ctx context.Context, m *sliceMachine, inv bigslice.Invocation, env CompileEnv) error {
-	// Substitute Result arguments for an invocation ref and record the
-	// dependency.
+func (b *bigmachineExecutor) compile(ctx context.Context, m *sliceMachine, inv execInvocation) error {
 	b.mu.Lock()
-	for i, arg := range inv.Args {
-		result, ok := arg.(*Result)
-		if !ok {
-			continue
-		}
-		inv.Args[i] = invocationRef{result.inv.Index}
-		if _, ok := b.invocations[result.inv.Index]; !ok {
-			b.mu.Unlock()
-			return fmt.Errorf("invalid result invocation %x", result.inv.Index)
+	if _, ok := b.invocations[inv.Index]; !ok {
+		// This is the first time we are seeing this invocation.
+
+		// Substitute each *Result argument for an invocationRef and record the
+		// dependency.
+		for i, arg := range inv.Args {
+			result, ok := arg.(*Result)
+			if !ok {
+				continue
+			}
+			inv.Args[i] = invocationRef{result.inv.Index}
+			if _, ok := b.invocations[result.inv.Index]; !ok {
+				b.mu.Unlock()
+				return fmt.Errorf("invalid result invocation %x", result.inv.Index)
+			}
+			if b.invocationDeps[inv.Index] == nil {
+				b.invocationDeps[inv.Index] = make(map[uint64]bool)
+			}
+			b.invocationDeps[inv.Index][result.inv.Index] = true
 		}
-		if b.invocationDeps[inv.Index] == nil {
-			b.invocationDeps[inv.Index] = make(map[uint64]bool)
+		b.invocations[inv.Index] = inv
+
+		// gob-encode the invocation, so we can reuse the work of gob-encoding
+		// when sending the invocation to each worker.
+		var invocationBuffer bytes.Buffer
+		enc := gob.NewEncoder(&invocationBuffer)
```

Comment:
Optional: Inline `enc` because it's short and only used once.

Reply:
Done.